drivers/soundwire/bus.c

```diff
@@ -1310,8 +1310,8 @@ static int sdw_handle_dp0_interrupt(struct sdw_slave *slave, u8 *slave_status)
 			*slave_status = clear;
 		}
 
-		/* clear the interrupt */
-		ret = sdw_write(slave, SDW_DP0_INT, clear);
+		/* clear the interrupt but don't touch reserved and SDCA_CASCADE fields */
+		ret = sdw_update(slave, SDW_DP0_INT, clear, clear);
```

Comment:
I don't have the data sheet, but the way this register is handled here, it *seems* to me, that this is a clear-on-write style register, isn't it? I.e. only bits that are set in the value being written, clear respective bits in the register, and bits, that are 0 don't change anything. Isn't this the case? If this is the case, then your change can potentially clear more unrelated bits?

Reply:
sdw_update is /* Read-Modify-Write Slave register */.  I don't get your idea.

Reply:
the SDCA_CASCADE bit is read-only and the reserved bits should not be touched as they may be reclaimed by MIPI for new usages. It makes no sense to force a value with a write. We should only update the bits that are defined by the spec and writable, and keep all other bits intact.

Reply:
> I don't have the data sheet, but the way this register is handled here, it _seems_ to me, that this is a clear-on-write style register, isn't it? I.e. only bits that are set in the value being written, clear respective bits in the register, and bits, that are 0 don't change anything. Isn't this the case? If this is the case, then your change can potentially clear more unrelated bits?

no, the other bits are not defined or not writable, so they are kept in whatever state they are precisely to avoid side effects and undocumented behavior.

Reply:
@plbossart @RanderWang let me explain. By the way `clear` is calculated and written to that register and from the comment, that register seems to be of the clear-on-set type. That means, say, if we have 3 interrupt bits: INT0 = BIT(0), INT1 = BIT(1), INT2 = BIT(2), and if we read 7 from the register, that means, that all three interrupt sources have triggered. If we write 1 to that register, BIT(0) will clear and if we re-read the register, it will have a value of 6. So, writing a *different* bit (we wrote 0 to BIT(1) and BIT(2)) doesn't change those values. Only writing 1 to a set bit clears it.
Now, with this change we would read-modify-write the register, i.e. we would read 7, modify BIT(0) to 1 (which it anyway already is), and write back 7 - thus clearing all 3 set interrupt bits and losing interrupt sources on BIT(1) and BIT(2). Even with currently reserved bits, if the register is of this kind, if other bits ever get used, they'll probably be of the same clear-on-set type. Or does this register work in a different way?

Reply:
> @plbossart @RanderWang let me explain. By the way `clear` is calculated and written to that register and from the comment, that register seems to be of the clear-on-set type. That means, say, if we have 3 interrupt bits: INT0 = BIT(0), INT1 = BIT(1), INT2 = BIT(2), and if we read 7 from the register, that means, that all three interrupt sources have triggered. If we write 1 to that register, BIT(0) will clear and if we re-read the register, it will have a value of 6. So, writing a _different_ bit (we wrote 0 to BIT(1) and BIT(2)) doesn't change those values. Only writing 1 to a set bit clears it.
> Now, with this change we would read-modify-write the register, i.e. we would read 7, modify BIT(0) to 1 (which it anyway already is), and write back 7 - thus clearing all 3 set interrupt bits and losing interrupt sources on BIT(1) and BIT(2). Even with currently reserved bits, if the register is of this kind, if other bits ever get used, they'll probably be of the same clear-on-set type. Or does this register work in a different way?

modify BIT(0) to 1,  why write back 7 ?  read_val & ~7 | 1

Reply:
@lyakh I also have no idea what the problem is. Yes the interrupt bits are clear on writes, and we only write1/clear them after handling them. The other bits are not write1/clear so we don't touch them and write the same value we read.

Reply:
@RanderWang @plbossart let's take an example. Let's say in line 1277 you read
```
status = sdw_read(slave, SDW_DP0_INT);
```
and at that time you read `status = 1`. In line 1287 you set `clear` to 1 too. Next you come to your *new* line 1314
```
ret = sdw_update(slave, SDW_DP0_INT, clear, clear);
```
and let's say now the same `SDW_DP0_INT` register already contains 7 because two more interrupts have just triggered. Now you do `sdw_write(slave, (7 & ~1) | 1);` i.e. you write 7 thus clearing all 3 interrupt sources.

Reply:
Thanks @lyakh I see your point. I guess we can force the 'clear' variable to mirror what the 'status' was and then clear the bits we need.

Reply:
Thanks, I remember this is the second time you mention this type of issue. But I don't remember why.  This is the issue of sdw_update.